File: Asgard_Test/tests_Heimdall/benchmarks/corpus/taint_js/fp_sa4_guard_clause_validator_clears.js
const { exec } = require('child_process');

// SA4 path-sensitivity: a classic early-return guard clause built on a
// REAL, catalog-whitelisted value-domain predicate (Number.isInteger --
// receiver-qualified, unshadowable). Every path that reaches the exec()
// call below has been proven to have a finite-number `host`, which cannot
// carry a shell-injection payload -- the guarded/validated path is clean.
app.post('/ping', (req, res) => {
    const host = req.body.host;
    if (!Number.isInteger(host)) {
        return res.status(400).send('invalid');
    }
    exec('ping -c 1 ' + host, (err, stdout) => {
        res.send(stdout);
    });
});

File: Asgard_Test/tests_Heimdall/benchmarks/corpus/taint/fp_sa4_guard_clause_validator_clears.py
"""FP: SA4 path-sensitivity -- an early-return guard clause built on a REAL
catalog-recognized value-domain predicate (`str.isdigit()`) proves every
path that reaches the sink has a digits-only `host`, which cannot carry a
shell-injection payload. The guarded/validated path is clean."""

import subprocess


def run():
    host = request.args.get("host")
    if not host.isdigit():
        return "invalid"
    subprocess.run("ping -c 1 " + host, shell=True)
